tools/c7n_mailer/c7n_mailer/gcp/pubsub_subscriptions.py

```diff
@@ -0,0 +1,184 @@
+        """Returns Rest API URI formatted with topic and project in it
+        """
+        return 'projects/{}/subscriptions/{}'.format(
+            project or session.get_default_project(),
+            subscription)
+
+    @staticmethod
+    def ensure_subscription(session, data, subscription=None):
+        """Verify the pub/sub topic exists.
+        If it does not, create it returns the topic qualified name.
+        """
+
+        client = session.client('pubsub', 'v1', 'projects.subscriptions')
+
+        topic = PubSubSubscriptions.ensure_topic(session, data)
+        subscription = PubSubSubscriptions.get_subscription_param(session, subscription)
+
+        try:
+            client.execute_command('get', {'subscription': subscription})
+        except HttpError as e:
+            if e.resp.status != 404:
+                raise
+        else:
+            return subscription
+
+        # bug in discovery doc.. apis say body must be empty but its required in the
+        # discovery api for create.
+        client.execute_command('create', {'name': subscription, 'body': {}})
+        return topic
+
+        pass
+
+    """
+    Class methods for handling Pub/Sub Topics
+    """
+    @staticmethod
+    def get_topic_param(session, data, topic=None, project=None):
+        """Returns Rest API URI formatted with topic and project in it
+        """
+        return 'projects/{}/topics/{}'.format(
+            project or session.get_default_project(),
+            topic or data['transport']['topic'])
+
+    @staticmethod
+    def ensure_topic(session, data):
+        """Verify the pub/sub topic exists.
+        If it does not, create it
+
+        Returns the topic qualified name.
+        """
+
+        client = session.client('pubsub', 'v1', 'projects.topics')
+
+        topic = PubSubSubscriptions.get_topic_param(session, data)
+        try:
+            client.execute_command('get', {'topic': topic})
+        except HttpError as e:
+            if e.resp.status != 404:
+                raise
+        else:
+            return topic
+
+        # bug in discovery doc.. apis say body must be empty but its required in the
+        # discovery api for create.
+        client.execute_command('create', {'name': topic, 'body': {}})
+        return topic
+
+    @staticmethod
+    def ensure_iam(session, data, publisher=None):
+        """ Ensures that the correct Iam Permissions are setup for a topic
+        """
+        topic = PubSubSubscriptions.get_topic_param(session, data)
+        client = session.client('pubsub', 'v1', 'projects.topics')
+        policy = client.execute_command('getIamPolicy', {'resource': topic})
+        policy.pop('etag')
+        found = False
+        for binding in policy.get('bindings', {}):
+            if binding['role'] != 'roles/pubsub.publisher':
+                continue
+            if publisher in binding['members']:
+                return
+            found = binding
+
+        if not found:
```

Comment:
I don't quite understand the context  of the for-loop above, but I have a feeling this conditional might not get triggered properly as is because of the way `found` is being set. Maybe this works because if this is meant to catch the condition where binding defaults to an `{}`, then it would hit this condiiton because `found = False` currently (it would still hit with `found = None` like I suggested changing above.) So if that's the thinking then I suppose it's fine